File: java-code/src/test/java/com/vd/study/java_code/container/ResultTest.java
package com.vd.study.java_code.container;

import org.junit.Assert;
import org.junit.Test;

public class ResultTest {

    @Test
    public void mapProgress_AfterMapping_ReturnNewProgressObject() {
        Result<Integer> result = new Result.Progress<>();

        Result<String> newObject = result.map(String::valueOf);

        Assert.assertTrue(newObject instanceof Result.Progress);
    }

    @Test(expected = Exception.class)
    public void mapError_AfterMapping_ReturnNewErrorObject() throws Exception {
        Result<Integer> result = new Result.Error<>(new IllegalArgumentException("EXCEPTION"));

        Result<String> newObject = result.map(String::valueOf);

        Assert.assertTrue(newObject instanceof Result.Error);
        Assert.assertNull(newObject.getOrNull());
        newObject.getOrException();
    }

    @Test
    public void mapCorrect_AfterMapping_ReturnNewCorrectObjectWithNewType() throws Exception {
        Result<Integer> result = new Result.Correct<>(12345);

        Result<String> newObject = result.map(String::valueOf);

        Assert.assertTrue(newObject instanceof Result.Correct);
        Assert.assertEquals("12345", newObject.getOrNull());
        Assert.assertEquals("12345", newObject.getOrException());
    }

    @Test
    public void getOrNullFunctionInCorrectComponent_AfterFunctionCalling_ReturnCorrectValue() {
        Result<String> correctObject = new Result.Correct<>("TEST");

        String result = correctObject.getOrNull();

        Assert.assertEquals("TEST", result);
    }

    @Test
    public void getOrExceptionFunctionInCorrectComponent_AfterFunctionCalling_ReturnCorrectValue() throws Exception {
        Result<String> correctObject = new Result.Correct<>("TEST");

        String result = correctObject.getOrException();

        Assert.assertEquals("TEST", result);
    }

    @Test
    public void getOrNullFunctionInProgressComponent_AfterFunctionCalling_ReturnNull() {
        Result<Void> progressObject = new Result.Progress<>();

        Void result = progressObject.getOrNull();

        Assert.assertNull(result);
    }

    @Test(expected = IllegalArgumentException.class)
    public void getOrExceptionFunctionInProgressComponent_AfterFunctionCalling_throwIllegalArgumentException() throws Exception {
        Result<Void> progressObject = new Result.Progress<>();

        progressObject.getOrException();
    }

    @Test
    public void getOrNullFunctionInErrorComponent_AfterFunctionCalling_ReturnNull() {
        Result<Void> errorObject = new Result.Error<>(new IllegalArgumentException());

        Void result = errorObject.getOrNull();

        Assert.assertNull(result);
    }

    @Test(expected = IllegalArgumentException.class)
    public void getOrExceptionFunctionInErrorComponent_AfterFunctionCalling_throwIllegalArgumentException() throws Exception {
        Result<Void> errorObject = new Result.Error<>(new IllegalArgumentException());

        errorObject.getOrException();
    }
}

File: kotlin-code/src/test/java/com/vd/study/kotlin_code/container/ResultTest.kt
package com.vd.study.kotlin_code.container

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ResultTest {

    @Test
    fun correctExceptionValue_AfterErrorObjectCreation_ContainsTheSameObject() {
        val exception = IllegalArgumentException("EXCEPTION")
        val errorObject = Result.Error(exception)

        assertEquals(exception, errorObject.exception)
    }

    @Test
    fun mapFunctionInProgressComponent_AfterFunctionCalling_ReturnTheSameObject() {
        val progressObject = Result.Progress

        val result = progressObject.map {}

        assertEquals(progressObject, result)
    }

    @Test
    fun mapFunctionInErrorComponent_AfterFunctionCalling_ReturnTheSameObject() {
        val exception = IllegalArgumentException()
        val errorObject = Result.Error(exception)

        val result = errorObject.map {}

        assertEquals(errorObject, result)
    }

    @Test
    fun mapFunctionInCorrectComponent_AfterFunctionCalling_ReturnObjectWithNewValueType() {
        val correctObject = Result.Correct("890")

        val result = correctObject.map { it.toInt() }

        assertEquals(890, result.getOrException())
    }

    @Test
    fun mapFunctionInCorrectComponent_AfterFunctionCalling_ReturnErrorComponent() {
        val correctObject = Result.Correct("TEST")

        val result = correctObject.map { it.toInt() }

        assertTrue(result is Result.Error)
    }

    @Test
    fun getOrNullFunctionInCorrectComponent_AfterFunctionCalling_ReturnCorrectValue() {
        val correctObject = Result.Correct("TEST")

        val result = correctObject.getOrNull()

        assertEquals("TEST", result)
    }

    @Test
    fun getOrExceptionFunctionInCorrectComponent_AfterFunctionCalling_ReturnCorrectValue() {
        val correctObject = Result.Correct("TEST")

        val result = correctObject.getOrException()

        assertEquals("TEST", result)
    }

    @Test
    fun getOrNullFunctionInProgressComponent_AfterFunctionCalling_ReturnNull() {
        val progressObject = Result.Progress

        val result = progressObject.getOrNull()

        assertNull(result)
    }

    @Test(expected = IllegalArgumentException::class)
    fun getOrExceptionFunctionInProgressComponent_AfterFunctionCalling_throwIllegalArgumentException() {
        val progressObject = Result.Progress

        progressObject.getOrException()
    }

    @Test
    fun getOrNullFunctionInErrorComponent_AfterFunctionCalling_ReturnNull() {
        val errorObject = Result.Error(IllegalArgumentException())

        val result = errorObject.getOrNull()

        assertNull(result)
    }

    @Test(expected = IllegalArgumentException::class)
    fun getOrExceptionFunctionInErrorComponent_AfterFunctionCalling_throwIllegalArgumentException() {
        val errorObject = Result.Error(IllegalArgumentException())

        errorObject.getOrException()
    }
}